File: jumble_solver.rb
#! /usr/bin/env ruby

require 'set'


if ARGV.size() < 1
    puts "usage: jumble_solver.rb DICIONARY_FILE [DICTIONARY_FILE] ..."
    exit(1)
end

sortedToOrigs = Hash.new()

ARGV.each { |dictFileName|
    IO.foreach(dictFileName) { |dictWord|
        dictWord.strip!()
        sortedWord = dictWord.split("").sort().join("")

        if !sortedToOrigs.has_key?(sortedWord)
            sortedToOrigs[sortedWord] = Set.new([])
        end

        sortedToOrigs[sortedWord].add(dictWord)
    }
}

print '$ '

while jumbledWord = STDIN.gets
    sortedWord = jumbledWord.strip!().split("").sort().join("")

    if sortedToOrigs.has_key?(sortedWord)
        puts sortedToOrigs[sortedWord].to_a().join(" ")
    else
        puts "no anagram in dictionary"
    end

    print '$ '
end

puts ""

File: jumble_solver.pl
#! /usr/bin/env perl

use strict;
use warnings;


if(scalar(@ARGV) < 1){
    print("usage: jumble_solver.pl DICTIONARY_FILE [DICTIONARY_FILE] ...\n");
}

my $sortedToOrigs = {};
for my $dictFileName (@ARGV) {
    open(DICT_FILE, "<$dictFileName");

    while(my $origWord = <DICT_FILE>){
        chomp($origWord);

        my $sortedWord = sortString($origWord);
        $sortedToOrigs->{$sortedWord}->{$origWord} = 1;
    }
    close(DICT_FILE);
}

print('$ ');

while(my $jumbledWord = <STDIN>){
    chomp($jumbledWord);
    my $sortedWord = sortString($jumbledWord);

    if(exists($sortedToOrigs->{$sortedWord})){
        print(join(" ", keys(%{$sortedToOrigs->{$sortedWord}})) . "\n");
    }
    else {
        print("no anagram found in dictionary\n");
    }

    print('$ ');
}
print("\n");


sub sortString {
    return join("", sort(split("",$_[0])));
}


